From b3efdd1f3a7d77fd2acca6787145b159404c3583 Mon Sep 17 00:00:00 2001
From: Arpit Agarwal <arp@apache.org>
Date: Fri, 18 Dec 2015 16:06:03 -0800
Subject: [PATCH 1046/2748] HADOOP-12656. MiniKdc throws "address in use"
 BindException. (Contributed by Wei-Chiu Chuang)

(cherry picked from commit 8fb70a031b323634ddc51ff6aff4f376baef68c8)

Change-Id: Ic96199dff33a793103fd4a849b2728396af87ac9
---
 .../java/org/apache/hadoop/minikdc/MiniKdc.java    |   22 +++++++++++---------
 1 file changed, 12 insertions(+), 10 deletions(-)

diff --git a/hadoop-common-project/hadoop-minikdc/src/main/java/org/apache/hadoop/minikdc/MiniKdc.java b/hadoop-common-project/hadoop-minikdc/src/main/java/org/apache/hadoop/minikdc/MiniKdc.java
index 4732283..23ef009 100644
--- a/hadoop-common-project/hadoop-minikdc/src/main/java/org/apache/hadoop/minikdc/MiniKdc.java
+++ b/hadoop-common-project/hadoop-minikdc/src/main/java/org/apache/hadoop/minikdc/MiniKdc.java
@@ -40,6 +40,7 @@
 import org.apache.directory.server.kerberos.shared.crypto.encryption.KerberosKeyFactory;
 import org.apache.directory.server.kerberos.shared.keytab.Keytab;
 import org.apache.directory.server.kerberos.shared.keytab.KeytabEntry;
+import org.apache.directory.server.protocol.shared.transport.AbstractTransport;
 import org.apache.directory.server.protocol.shared.transport.TcpTransport;
 import org.apache.directory.server.protocol.shared.transport.UdpTransport;
 import org.apache.directory.server.xdbm.Index;
@@ -63,8 +64,7 @@
 import java.io.IOException;
 import java.io.StringReader;
 import java.lang.reflect.Method;
-import java.net.InetAddress;
-import java.net.ServerSocket;
+import java.net.InetSocketAddress;
 import java.text.MessageFormat;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -265,12 +265,6 @@ public MiniKdc(Properties conf, File workDir) throws Exception {
     LOG.info("---------------------------------------------------------------");
     this.conf = conf;
     port = Integer.parseInt(conf.getProperty(KDC_PORT));
-    if (port == 0) {
-      ServerSocket ss = new ServerSocket(0, 1, InetAddress.getByName
-              (conf.getProperty(KDC_BIND_ADDRESS)));
-      port = ss.getLocalPort();
-      ss.close();
-    }
     String orgName= conf.getProperty(ORG_NAME);
     String orgDomain = conf.getProperty(ORG_DOMAIN);
     realm = orgName.toUpperCase(Locale.ENGLISH) + "."
@@ -450,13 +444,15 @@ private void initKDCServer() throws Exception {
 
     // transport
     String transport = conf.getProperty(TRANSPORT);
+    AbstractTransport absTransport;
     if (transport.trim().equals("TCP")) {
-      kdc.addTransports(new TcpTransport(bindAddress, port, 3, 50));
+      absTransport = new TcpTransport(bindAddress, port, 3, 50);
     } else if (transport.trim().equals("UDP")) {
-      kdc.addTransports(new UdpTransport(port));
+      absTransport = new UdpTransport(port);
     } else {
       throw new IllegalArgumentException("Invalid transport: " + transport);
     }
+    kdc.addTransports(absTransport);
     kdc.setServiceName(conf.getProperty(INSTANCE));
     kdc.getConfig().setMaximumRenewableLifetime(
             Long.parseLong(conf.getProperty(MAX_RENEWABLE_LIFETIME)));
@@ -465,6 +461,12 @@ private void initKDCServer() throws Exception {
 
     kdc.getConfig().setPaEncTimestampRequired(false);
     kdc.start();
+    // if using ephemeral port, update port number for binding
+    if (port == 0) {
+      InetSocketAddress addr =
+          (InetSocketAddress)absTransport.getAcceptor().getLocalAddress();
+      port = addr.getPort();
+    }
 
     StringBuilder sb = new StringBuilder();
     InputStream is2 = getResourceAsStream("minikdc-krb5.conf");
-- 
1.7.9.5

